python/pyspark/sql/types.py

```diff
@@ -64,6 +64,8 @@ def json(self):
                           sort_keys=True)
 
 
+# This singleton pattern does not work with pickle, you will get
```

Comment:
We discussed this, the old implementation is a bit of over optimization, I think.

For PrimitiveTypeSingleton, the __dict__ is empty, so the new implementation is not so heavy, so there will be no much performance difference.